File: app/macos/hyperwhisper/Utilities/AppClassification/AppTypeClassifier.swift
//
//  AppTypeClassifier.swift
//  hyperwhisper
//
//  Shared catalog-backed application classification for app-aware formatting.
//
//  The ALGORITHM is not here. Issue #279 moved it into `hw-catalog` and this is
//  now a facade over `appClassify` / `appIsWebmail`: the 8-element priority
//  array, the keyword-prep rule, the word-boundary rule, the host-suffix rule
//  and the email regex existed in Swift, C# and Rust, and had already drifted.
//  `shared-conformance/app-type-vectors.json`, run by
//  `AppTypeConformanceVectorTests`, pins the answer this file returns.
//

import Foundation

public enum AppType: String, Codable {
    case email
    case ai
    case workMessaging
    case personalMessaging
    case document
    case code
    case terminal
    case sensitive
    case other

    var promptValue: String {
        switch self {
        case .workMessaging:
            return "work_messaging"
        case .personalMessaging:
            return "personal_messaging"
        default:
            return rawValue
        }
    }

    var category: String {
        switch self {
        case .email:
            return "Email Client"
        case .ai:
            return "AI"
        case .workMessaging, .personalMessaging:
            return "Communication"
        case .document:
            return "Document"
        case .code:
            return "Code Editor"
        case .terminal:
            return "Terminal"
        case .sensitive:
            return "Sensitive"
        case .other:
            return "Application"
        }
    }

    var textInputFormat: String {
        switch self {
        case .email:
            return "email"
        case .code:
            return "code"
        case .terminal:
            return "command"
        case .document:
            return "markdown"
        default:
            return "text"
        }
    }

    fileprivate init(_ classified: ClassifiedAppType) {
        switch classified {
        case .email: self = .email
        case .ai: self = .ai
        case .workMessaging: self = .workMessaging
        case .personalMessaging: self = .personalMessaging
        case .document: self = .document
        case .code: self = .code
        case .terminal: self = .terminal
        case .sensitive: self = .sensitive
        case .other: self = .other
        }
    }
}

public struct AppClassificationResult {
    let appType: AppType
    let confidence: String
    let source: String
    let matched: String?
}

public final class AppTypeClassifier {
    public static let shared = AppTypeClassifier()

    private init() {}

    /// Classify the frontmost app. Signals are tried in order — host, bundle
    /// id, title, app name, focused element — and the first hit wins.
    ///
    /// `browserTitle` is the browser TAB title, not the window title. Windows
    /// joins both; the shared core takes an already-composed string, so
    /// widening this is a change here and nowhere else.
    public func classify(
        bundleId: String,
        appName: String,
        browserHost: String?,
        browserTitle: String?,
        focusedElement: FocusedElementInfo?
    ) -> AppClassificationResult {
        let result = appClassify(request: AppClassifyRequest(
            bundleId: bundleId.trimmingCharacters(in: .whitespacesAndNewlines),
            processName: "",
            appName: appName,
            host: browserHost,
            // macOS has always reported a host hit as `strong`. Windows reports
            // `medium` and the Local API `manual`, and that string reaches the
            // LLM prompt, so the caller owns it rather than the core.
            hostConfidence: "strong",
            title: browserTitle ?? "",
            focusedPieces: Self.focusedPieces(focusedElement)
        ))

        return AppClassificationResult(
            appType: AppType(result.appType),
            confidence: result.confidence,
            source: result.source,
            matched: result.matched
        )
    }

    /// Whether a browser-tab title looks like webmail. Call this ONLY when the
    /// frontmost app is already known to be a browser and nothing else
    /// classified the window — a title is not evidence of webmail on its own.
    public static func isWebmail(_ tabTitle: String) -> Bool {
        appIsWebmail(title: tabTitle)
    }

    /// The five accessibility fields macOS reads. Windows supplies two; the
    /// core takes whatever the platform has, joins the non-blank pieces and
    /// scans the result.
    ///
    /// These values are NOT truncated on the way in. `PromptBuilder` bounds
    /// `focusedContent` to 100 characters before it reaches the prompt, but the
    /// email scan has always run against the full value, and truncating first
    /// would drop an address that sits past the cut.
    private static func focusedPieces(_ focusedElement: FocusedElementInfo?) -> [String] {
        guard let focusedElement else { return [] }
        return [
            focusedElement.role,
            focusedElement.title,
            focusedElement.description,
            focusedElement.placeholder,
            focusedElement.value
        ].compactMap { $0 }
    }
}

File: app/shared-dotnet/HyperWhisper.AppClassification/AppTypeClassifier.cs
using uniffi.hyperwhisper_core;

namespace HyperWhisper.AppClassification;

/// <summary>
/// Coarse application type driving app-aware formatting.
/// </summary>
public enum AppType
{
    Email,
    Ai,
    WorkMessaging,
    PersonalMessaging,
    Document,
    Code,
    Terminal,
    Sensitive,
    Other
}

/// <summary>
/// The derived strings for an <see cref="AppType"/>. The shared core resolves
/// the same three strings and returns them on every classification, and
/// <c>HyperWhisper.AppTypeConformance.Tests</c> asserts these switches against
/// that answer, so the pair cannot drift.
/// </summary>
public static class AppTypeExtensions
{
    public static string ToPromptValue(this AppType appType) => appType switch
    {
        AppType.WorkMessaging => "work_messaging",
        AppType.PersonalMessaging => "personal_messaging",
        _ => appType.ToString().ToLowerInvariant()
    };

    public static string ToCategory(this AppType appType) => appType switch
    {
        AppType.Email => "Email Client",
        AppType.Ai => "AI",
        AppType.WorkMessaging or AppType.PersonalMessaging => "Communication",
        AppType.Document => "Document",
        AppType.Code => "Code Editor",
        AppType.Terminal => "Terminal",
        AppType.Sensitive => "Sensitive",
        _ => "Application"
    };

    public static string ToTextFormat(this AppType appType) => appType switch
    {
        AppType.Email => "email",
        AppType.Code => "code",
        AppType.Terminal => "command",
        AppType.Document => "markdown",
        _ => "text"
    };
}

/// <summary>
/// Everything the platform observed about the foreground app. Pass an empty
/// string, <c>null</c>, or an empty list for a signal this platform cannot see.
/// </summary>
/// <param name="BundleId">macOS bundle identifier. Always empty on .NET heads.</param>
/// <param name="ProcessName">Process name without an extension, e.g. <c>OUTLOOK</c> or <c>konsole</c>.</param>
/// <param name="AppName">The app's display name, when the platform has one distinct from the process.</param>
/// <param name="Host">Browser host for a web app. A full URL is accepted and normalized.</param>
/// <param name="HostConfidence">
/// The confidence to report for a host hit. It reaches the LLM prompt, so the
/// caller owns it; an empty value means <c>strong</c>.
/// </param>
/// <param name="Title">Window and/or browser-tab title, composed by the caller.</param>
/// <param name="FocusedPieces">Text read off the focused accessibility element.</param>
public sealed record AppClassificationRequest(
    string BundleId = "",
    string ProcessName = "",
    string AppName = "",
    string? Host = null,
    string HostConfidence = "",
    string Title = "",
    IReadOnlyList<string>? FocusedPieces = null);

public sealed record AppClassificationResult(
    AppType AppType,
    string Confidence,
    string Source,
    string? Matched);

/// <summary>
/// Catalog-backed application classification, for every .NET head.
/// </summary>
/// <remarks>
/// The algorithm itself lives in <c>hw-catalog</c> and nowhere else — issue
/// #279 deleted the 320-line C# copy and the 300-line Swift one. What is left
/// here is the mapping between the generated binding's types and the platform's
/// own, which is why the classifier can now be reached from Linux at all.
/// </remarks>
public static class AppTypeClassifier
{
    /// <summary>
    /// Classify the foreground app. Signals are tried in order — host, bundle
    /// id, process name, title, app name, focused element — and the first hit
    /// wins; otherwise the result is <see cref="AppType.Other"/> at
    /// <c>unknown</c>/<c>default</c>.
    /// </summary>
    public static AppClassificationResult Classify(AppClassificationRequest request)
    {
        ArgumentNullException.ThrowIfNull(request);

        var classification = HyperwhisperCoreMethods.AppClassify(new AppClassifyRequest(
            @bundleId: request.BundleId ?? string.Empty,
            @processName: request.ProcessName ?? string.Empty,
            @appName: request.AppName ?? string.Empty,
            @host: string.IsNullOrWhiteSpace(request.Host) ? null : request.Host,
            @hostConfidence: request.HostConfidence ?? string.Empty,
            @title: request.Title ?? string.Empty,
            @focusedPieces: request.FocusedPieces is null
                ? []
                : [.. request.FocusedPieces.Where(piece => !string.IsNullOrWhiteSpace(piece))]));

        return new AppClassificationResult(
            FromBinding(classification.@appType),
            classification.@confidence,
            classification.@source,
            classification.@matched);
    }

    /// <summary>
    /// Whether a browser-tab title looks like webmail. Call this ONLY when the
    /// foreground app is already known to be a browser and nothing else
    /// classified the window — a title is not evidence of webmail on its own.
    /// </summary>
    public static bool IsWebmail(string title)
    {
        ArgumentNullException.ThrowIfNull(title);
        return HyperwhisperCoreMethods.AppIsWebmail(title);
    }

    private static AppType FromBinding(ClassifiedAppType appType) => appType switch
    {
        ClassifiedAppType.Email => AppType.Email,
        ClassifiedAppType.Ai => AppType.Ai,
        ClassifiedAppType.WorkMessaging => AppType.WorkMessaging,
        ClassifiedAppType.PersonalMessaging => AppType.PersonalMessaging,
        ClassifiedAppType.Document => AppType.Document,
        ClassifiedAppType.Code => AppType.Code,
        ClassifiedAppType.Terminal => AppType.Terminal,
        ClassifiedAppType.Sensitive => AppType.Sensitive,
        ClassifiedAppType.Other => AppType.Other,
        _ => AppType.Other
    };
}
